File: scripts/test_system.py
#!/usr/bin/env python3

"""
Script de teste para validar as principais funcionalidades do sistema CRM
Execute com: python3 scripts/test_system.py
"""

import os
import sys
from pathlib import Path

# Cores para output
class Colors:
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    BOLD = '\033[1m'
    RESET = '\033[0m'

def log(message, color=Colors.RESET):
    print(f"{color}{message}{Colors.RESET}")

def log_success(message):
    log(f"✅ {message}", Colors.GREEN)

def log_error(message):
    log(f"❌ {message}", Colors.RED)

def log_warning(message):
    log(f"⚠️  {message}", Colors.YELLOW)

def log_info(message):
    log(f"ℹ️  {message}", Colors.BLUE)

def file_exists(file_path):
    return Path(file_path).exists()

def dir_exists(dir_path):
    return Path(dir_path).is_dir()

def check_file_content(file_path, search_string):
    if not file_exists(file_path):
        return False
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            return search_string in content
    except:
        return False

def test_project_structure():
    log(f"\n📁 Testando estrutura do projeto...", Colors.BOLD)
    
    required_dirs = [
        'app/api',
        'components',
        'lib',
        'hooks',
        'types',
        'middleware',
        'scripts'
    ]
    
    passed = 0
    
    for dir_path in required_dirs:
        if dir_exists(dir_path):
            log_success(f"Diretório {dir_path} existe")
            passed += 1
        else:
            log_error(f"Diretório {dir_path} não encontrado")
    
    return passed == len(required_dirs)

def test_api_routes():
    log(f"\n🔌 Testando rotas da API...", Colors.BOLD)
    
    api_routes = [
        'app/api/comercial/oportunidades/route.ts',
        'app/api/comercial/oportunidades/[id]/route.ts',
        'app/api/comercial/oportunidades/[id]/documentos/route.ts',
        'app/api/licitacoes/route.ts',
        'app/api/licitacoes/[id]/route.ts',
        'app/api/documentos/route.ts',
        'app/api/documentos/por-oportunidade/route.ts',
        'app/api/documentos/por-tag/route.ts',
        'app/api/contatos/route.ts',
        'app/api/contatos/[id]/route.ts'
    ]
    
    passed = 0
    
    for route in api_routes:
        if file_exists(route):
            log_success(f"Rota {route} existe")
            passed += 1
        else:
            log_error(f"Rota {route} não encontrada")
    
    return passed == len(api_routes)

def test_components():
    log(f"\n🧩 Testando componentes...", Colors.BOLD)
    
    components = [
        'components/detalhes-oportunidade.tsx',
        'components/licitacoes/detalhes-licitacao.tsx',
        'components/detalhes-cliente.tsx',
        'components/licitacoes/detalhes-orgao.tsx',
        'components/documentos/visualizador-documentos.tsx'
    ]
    
    passed = 0
    
    for component in components:
        if file_exists(component):
            log_success(f"Componente {component} existe")
            passed += 1
        else:
            log_error(f"Componente {component} não encontrado")
    
    return passed == len(components)

def test_security():
    log(f"\n🔒 Testando implementações de segurança...", Colors.BOLD)
    
    security_files = [
        'middleware/security.ts',
        'lib/rate-limit.ts',
        'lib/cache.ts'
    ]
    
    passed = 0
    
    for file_path in security_files:
        if file_exists(file_path):
            log_success(f"Arquivo de segurança {file_path} existe")
            passed += 1
        else:
            log_error(f"Arquivo de segurança {file_path} não encontrado")
    
    # Verificar se middleware de segurança tem funções essenciais
    if file_exists('middleware/security.ts'):
        security_functions = [
            'sanitizeInput',
            'isValidUUID',
            'isValidEmail',
            'isValidCNPJ',
            'securityMiddleware'
        ]
        
        for func in security_functions:
            if check_file_content('middleware/security.ts', func):
                log_success(f"Função de segurança {func} implementada")
                passed += 1
            else:
                log_error(f"Função de segurança {func} não encontrada")
    
    return passed >= len(security_files)

def test_sql_fixes():
    log(f"\n🗄️  Testando correções SQL...", Colors.BOLD)
    
    sql_file = 'sql_fixes.sql'
    
    if not file_exists(sql_file):
        log_error(f"Arquivo {sql_file} não encontrado")
        return False
    
    sql_checks = [
        'ALTER TABLE documentos',
        'FOREIGN KEY',
        'CREATE INDEX',
        'CREATE TRIGGER',
        'CREATE FUNCTION fn_validar_cnpj',
        'CREATE OR REPLACE VIEW view_estatisticas_sistema',
        'CREATE PROCEDURE sp_limpeza_dados_orfaos'
    ]
    
    passed = 0
    
    for check in sql_checks:
        if check_file_content(sql_file, check):
            log_success(f"Correção SQL encontrada: {check}")
            passed += 1
        else:
            log_warning(f"Correção SQL não encontrada: {check}")
    
    return passed >= len(sql_checks) * 0.8  # 80% das correções devem estar presentes

def test_documentation():
    log(f"\n📚 Testando documentação...", Colors.BOLD)
    
    doc_files = [
        'todo.md',
        'melhorias_robustez.md',
        'sql_fixes.sql'
    ]
    
    passed = 0
    
    for file_path in doc_files:
        if file_exists(file_path):
            log_success(f"Arquivo de documentação {file_path} existe")
            passed += 1
        else:
            log_error(f"Arquivo de documentação {file_path} não encontrado")
    
    return passed == len(doc_files)

def test_tag_system():
    log(f"\n🏷️  Testando sistema de tags...", Colors.BOLD)
    
    passed = 0
    
    # Verificar se a nova rota de documentos por oportunidade existe
    if file_exists('app/api/comercial/oportunidades/[id]/documentos/route.ts'):
        log_success('Nova rota de documentos por oportunidade implementada')
        passed += 1
    else:
        log_error('Nova rota de documentos por oportunidade não encontrada')
    
    # Verificar se a correção na API de oportunidades foi feita
    if check_file_content('app/api/comercial/oportunidades/route.ts', 'documentos_vinculados'):
        log_success('Correção do sistema de tags na API de oportunidades implementada')
        passed += 1
    else:
        log_warning('Correção do sistema de tags pode não estar completa')
    
    return passed >= 1

def test_document_viewer():
    log(f"\n📄 Testando visualizador de documentos...", Colors.BOLD)
    
    passed = 0
    
    if file_exists('components/documentos/visualizador-documentos.tsx'):
        log_success('Componente VisualizadorDocumentos existe')
        passed += 1
        
        features = [
            'entityType',
            'showFilters',
            'handlePreview',
            'handleDownload',
            'formatarTamanho'
        ]
        
        for feature in features:
            if check_file_content('components/documentos/visualizador-documentos.tsx', feature):
                log_success(f"Funcionalidade {feature} implementada")
                passed += 1
            else:
                log_warning(f"Funcionalidade {feature} pode estar ausente")
    else:
        log_error('Componente VisualizadorDocumentos não encontrado')
    
    return passed >= 3

def run_tests():
    log('🚀 Iniciando testes do sistema CRM...', Colors.BOLD)
    log('=' * 50)
    
    tests = [
        {'name': 'Estrutura do Projeto', 'fn': test_project_structure},
        {'name': 'Rotas da API', 'fn': test_api_routes},
        {'name': 'Componentes', 'fn': test_components},
        {'name': 'Segurança', 'fn': test_security},
        {'name': 'Correções SQL', 'fn': test_sql_fixes},
        {'name': 'Documentação', 'fn': test_documentation},
        {'name': 'Sistema de Tags', 'fn': test_tag_system},
        {'name': 'Visualizador de Documentos', 'fn': test_document_viewer}
    ]
    
    total_passed = 0
    results = []
    
    for test in tests:
        passed = test['fn']()
        results.append({'name': test['name'], 'passed': passed})
        if passed:
            total_passed += 1
    
    # Resumo dos resultados
    log(f"\n📊 Resumo dos Testes:", Colors.BOLD)
    log('=' * 50)
    
    for result in results:
        if result['passed']:
            log_success(f"{result['name']}: PASSOU")
        else:
            log_error(f"{result['name']}: FALHOU")
    
    log(f"\n📈 Resultado Final: {total_passed}/{len(tests)} testes passaram", Colors.BOLD)
    
    if total_passed == len(tests):
        log_success('🎉 Todos os testes passaram! Sistema está pronto para produção.')
    elif total_passed >= len(tests) * 0.8:
        log_warning('⚠️  A maioria dos testes passou. Verifique os itens que falharam.')
    else:
        log_error('❌ Muitos testes falharam. Sistema precisa de mais correções.')
    
    return total_passed / len(tests)

if __name__ == '__main__':
    score = run_tests()
    sys.exit(0 if score == 1 else 1)

File: scripts/test-system.js
#!/usr/bin/env node

/**
 * Script de teste para validar as principais funcionalidades do sistema CRM
 * Execute com: node scripts/test-system.js
 */

const fs = require('fs');
const path = require('path');

// Cores para output
const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m',
  bold: '\x1b[1m'
};

function log(message, color = 'reset') {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

function logSuccess(message) {
  log(`✅ ${message}`, 'green');
}

function logError(message) {
  log(`❌ ${message}`, 'red');
}

function logWarning(message) {
  log(`⚠️  ${message}`, 'yellow');
}

function logInfo(message) {
  log(`ℹ️  ${message}`, 'blue');
}

// Verificar se arquivo existe
function fileExists(filePath) {
  return fs.existsSync(filePath);
}

// Verificar se diretório existe
function dirExists(dirPath) {
  return fs.existsSync(dirPath) && fs.statSync(dirPath).isDirectory();
}

// Verificar conteúdo do arquivo
function checkFileContent(filePath, searchString) {
  if (!fileExists(filePath)) return false;
  const content = fs.readFileSync(filePath, 'utf8');
  return content.includes(searchString);
}

// Testes de estrutura do projeto
function testProjectStructure() {
  log('\n📁 Testando estrutura do projeto...', 'bold');
  
  const requiredDirs = [
    'app/api',
    'components',
    'lib',
    'hooks',
    'types',
    'middleware',
    'scripts'
  ];
  
  let passed = 0;
  
  requiredDirs.forEach(dir => {
    if (dirExists(dir)) {
      logSuccess(`Diretório ${dir} existe`);
      passed++;
    } else {
      logError(`Diretório ${dir} não encontrado`);
    }
  });
  
  return passed === requiredDirs.length;
}

// Testes de APIs
function testAPIRoutes() {
  log('\n🔌 Testando rotas da API...', 'bold');
  
  const apiRoutes = [
    'app/api/comercial/oportunidades/route.ts',
    'app/api/comercial/oportunidades/[id]/route.ts',
    'app/api/comercial/oportunidades/[id]/documentos/route.ts',
    'app/api/licitacoes/route.ts',
    'app/api/licitacoes/[id]/route.ts',
    'app/api/documentos/route.ts',
    'app/api/documentos/por-oportunidade/route.ts',
    'app/api/documentos/por-tag/route.ts',
    'app/api/contatos/route.ts',
    'app/api/contatos/[id]/route.ts'
  ];
  
  let passed = 0;
  
  apiRoutes.forEach(route => {
    if (fileExists(route)) {
      logSuccess(`Rota ${route} existe`);
      passed++;
    } else {
      logError(`Rota ${route} não encontrada`);
    }
  });
  
  return passed === apiRoutes.length;
}

// Testes de componentes
function testComponents() {
  log('\n🧩 Testando componentes...', 'bold');
  
  const components = [
    'components/detalhes-oportunidade.tsx',
    'components/licitacoes/detalhes-licitacao.tsx',
    'components/detalhes-cliente.tsx',
    'components/licitacoes/detalhes-orgao.tsx',
    'components/documentos/visualizador-documentos.tsx'
  ];
  
  let passed = 0;
  
  components.forEach(component => {
    if (fileExists(component)) {
      logSuccess(`Componente ${component} existe`);
      passed++;
    } else {
      logError(`Componente ${component} não encontrado`);
    }
  });
  
  return passed === components.length;
}

// Testes de middleware e segurança
function testSecurity() {
  log('\n🔒 Testando implementações de segurança...', 'bold');
  
  const securityFiles = [
    'middleware/security.ts',
    'lib/rate-limit.ts',
    'lib/cache.ts'
  ];
  
  let passed = 0;
  
  securityFiles.forEach(file => {
    if (fileExists(file)) {
      logSuccess(`Arquivo de segurança ${file} existe`);
      passed++;
    } else {
      logError(`Arquivo de segurança ${file} não encontrado`);
    }
  });
  
  // Verificar se middleware de segurança tem funções essenciais
  if (fileExists('middleware/security.ts')) {
    const securityFunctions = [
      'sanitizeInput',
      'isValidUUID',
      'isValidEmail',
      'isValidCNPJ',
      'securityMiddleware'
    ];
    
    securityFunctions.forEach(func => {
      if (checkFileContent('middleware/security.ts', func)) {
        logSuccess(`Função de segurança ${func} implementada`);
        passed++;
      } else {
        logError(`Função de segurança ${func} não encontrada`);
      }
    });
  }
  
  return passed >= securityFiles.length;
}

// Testes de correções SQL
function testSQLFixes() {
  log('\n🗄️  Testando correções SQL...', 'bold');
  
  const sqlFile = 'sql_fixes.sql';
  
  if (!fileExists(sqlFile)) {
    logError(`Arquivo ${sqlFile} não encontrado`);
    return false;
  }
  
  const sqlChecks = [
    'ALTER TABLE documentos',
    'FOREIGN KEY',
    'CREATE INDEX',
    'CREATE TRIGGER',
    'CREATE FUNCTION fn_validar_cnpj',
    'CREATE OR REPLACE VIEW view_estatisticas_sistema',
    'CREATE PROCEDURE sp_limpeza_dados_orfaos'
  ];
  
  let passed = 0;
  
  sqlChecks.forEach(check => {
    if (checkFileContent(sqlFile, check)) {
      logSuccess(`Correção SQL encontrada: ${check}`);
      passed++;
    } else {
      logWarning(`Correção SQL não encontrada: ${check}`);
    }
  });
  
  return passed >= sqlChecks.length * 0.8; // 80% das correções devem estar presentes
}

// Testes de documentação
function testDocumentation() {
  log('\n📚 Testando documentação...', 'bold');
  
  const docFiles = [
    'todo.md',
    'melhorias_robustez.md',
    'sql_fixes.sql'
  ];
  
  let passed = 0;
  
  docFiles.forEach(file => {
    if (fileExists(file)) {
      logSuccess(`Arquivo de documentação ${file} existe`);
      passed++;
    } else {
      logError(`Arquivo de documentação ${file} não encontrado`);
    }
  });
  
  return passed === docFiles.length;
}

// Verificar integridade do sistema de tags
function testTagSystem() {
  log('\n🏷️  Testando sistema de tags...', 'bold');
  
  let passed = 0;
  
  // Verificar se a nova rota de documentos por oportunidade existe
  if (fileExists('app/api/comercial/oportunidades/[id]/documentos/route.ts')) {
    logSuccess('Nova rota de documentos por oportunidade implementada');
    passed++;
  } else {
    logError('Nova rota de documentos por oportunidade não encontrada');
  }
  
  // Verificar se a correção na API de oportunidades foi feita
  if (checkFileContent('app/api/comercial/oportunidades/route.ts', 'documentos_vinculados')) {
    logSuccess('Correção do sistema de tags na API de oportunidades implementada');
    passed++;
  } else {
    logWarning('Correção do sistema de tags pode não estar completa');
  }
  
  return passed >= 1;
}

// Verificar visualizador de documentos
function testDocumentViewer() {
  log('\n📄 Testando visualizador de documentos...', 'bold');
  
  let passed = 0;
  
  if (fileExists('components/documentos/visualizador-documentos.tsx')) {
    logSuccess('Componente VisualizadorDocumentos existe');
    passed++;
    
    const features = [
      'entityType',
      'showFilters',
      'handlePreview',
      'handleDownload',
      'formatarTamanho'
    ];
    
    features.forEach(feature => {
      if (checkFileContent('components/documentos/visualizador-documentos.tsx', feature)) {
        logSuccess(`Funcionalidade ${feature} implementada`);
        passed++;
      } else {
        logWarning(`Funcionalidade ${feature} pode estar ausente`);
      }
    });
  } else {
    logError('Componente VisualizadorDocumentos não encontrado');
  }
  
  return passed >= 3;
}

// Função principal de teste
function runTests() {
  log('🚀 Iniciando testes do sistema CRM...', 'bold');
  log('=' * 50);
  
  const tests = [
    { name: 'Estrutura do Projeto', fn: testProjectStructure },
    { name: 'Rotas da API', fn: testAPIRoutes },
    { name: 'Componentes', fn: testComponents },
    { name: 'Segurança', fn: testSecurity },
    { name: 'Correções SQL', fn: testSQLFixes },
    { name: 'Documentação', fn: testDocumentation },
    { name: 'Sistema de Tags', fn: testTagSystem },
    { name: 'Visualizador de Documentos', fn: testDocumentViewer }
  ];
  
  let totalPassed = 0;
  const results = [];
  
  tests.forEach(test => {
    const passed = test.fn();
    results.push({ name: test.name, passed });
    if (passed) totalPassed++;
  });
  
  // Resumo dos resultados
  log('\n📊 Resumo dos Testes:', 'bold');
  log('=' * 50);
  
  results.forEach(result => {
    if (result.passed) {
      logSuccess(`${result.name}: PASSOU`);
    } else {
      logError(`${result.name}: FALHOU`);
    }
  });
  
  log(`\n📈 Resultado Final: ${totalPassed}/${tests.length} testes passaram`, 'bold');
  
  if (totalPassed === tests.length) {
    logSuccess('🎉 Todos os testes passaram! Sistema está pronto para produção.');
  } else if (totalPassed >= tests.length * 0.8) {
    logWarning('⚠️  A maioria dos testes passou. Verifique os itens que falharam.');
  } else {
    logError('❌ Muitos testes falharam. Sistema precisa de mais correções.');
  }
  
  return totalPassed / tests.length;
}

// Executar testes se o script for chamado diretamente
if (require.main === module) {
  const score = runTests();
  process.exit(score === 1 ? 0 : 1);
}

module.exports = { runTests };

